test/JoinTest.cpp

```diff
@@ -0,0 +1,429 @@
+
+#include <cstdio>
+#include <fstream>
+#include <sstream>
+#include <algorithm>
+#include <tuple>
+
+#include "engine/CallFixedSize.h"
+#include "engine/Engine.h"
+#include "engine/Join.h"
+#include "engine/OptionalJoin.h"
+#include "util/Random.h"
+#include "engine/QueryExecutionTree.h"
+#include "util/Forward.h"
+#include "util/SourceLocation.h"
+#include "./util/GTestHelpers.h"
+#include "engine/idTable/IdTable.h"
+
+ad_utility::AllocatorWithLimit<Id>& allocator() {
+  static ad_utility::AllocatorWithLimit<Id> a{
+      ad_utility::makeAllocationMemoryLeftThreadsafeObject(
+          std::numeric_limits<size_t>::max())};
+  return a;
+}
+
+auto I = [](const auto& id) {
+  return Id::makeFromVocabIndex(VocabIndex::make(id));
+};
+
+// For easier reading. We repeat that type combination so often, that this
+// will make things a lot easier in terms of reading and writing.
+using vectorTable = std::vector<std::vector<size_t>>;
+
+/*
+ * Return an 'IdTable' with the given 'tableContent'. all rows must have the
+ * same length.
+*/
+IdTable makeIdTableFromVector(vectorTable tableContent) {
+  AD_CHECK(!tableContent.empty());
+  IdTable result{tableContent[0].size(), allocator()};
+
+  // Copying the content into the table.
+  for (const auto& row: tableContent) {
+    AD_CHECK(row.size() == result.numColumns()); // All rows of an IdTable must have the same length.
+    const size_t backIndex{result.size()};
+    result.emplace_back();
+
+    for (size_t c = 0; c < row.size(); c++) {
+      result(backIndex, c) = I(row[c]);
+    }
+  }
+
+  return result;
+}
+
+/*
+ * Does what it says on the tin: Save an IdTable with the corresponding
+ * join column.
+ */
+struct IdTableAndJoinColumn {
+  IdTable idTable;
+  size_t joinColumn;
+};
+
+/*
+ * A structure containing all information needed for a normal join test. A
+ * normal join test is defined as two IdTables being joined, the resulting
+ * IdTable tested, if it is sorted after the join column, or not, and this
+ * IdTable is than compared with the expectedResult.
+ */
+struct normalJoinTest {
+  IdTableAndJoinColumn leftInput;
+  IdTableAndJoinColumn rightInput;
+  IdTable expectedResult;
+  bool resultMustBeSortedByJoinColumn;
+};
+
+/*
+ * @brief Tests, whether the given IdTable has the same content as the sample
+ * solution and, if the option was choosen, if the IdTable is sorted by
+ * the join column.
+ *
+ * @param table The IdTable that should be tested.
+ * @param expectedContent The sample solution. Doesn't need to be sorted,
+ *  or the same order of rows as the table.
+ * @param resultMustBeSortedByJoinColumn If this is true, it will also be tested, if the table
+ *  is sorted by the join column.
+ * @param joinColumn The join column of the table.
+ * @param l Ignore it. It's only here for being able to make better messages,
+ *  if a IdTable fails the comparison.
+*/
+void compareIdTableWithExpectedContent(const IdTable& table, 
+    const IdTable& expectedContent,
```

Comment:
I actually did it like that, so that the user doesn't have to call `std::move`, or `clone()`.

There is no real easy way to signal, that you need to use `std::move()`, or `clone()`, without reading the function documentation, which most people will no do closely enough, to see that. Combined with how unclear the compiler error message is for a deleted copy constructor, most people would fight with a call of this function for a while.

By using an alias and copying it localy, that problem vanishes and it's more user friendly.